doc/developer/design/20231113_optimizer_notice_catalog.md

```diff
@@ -0,0 +1,555 @@
+  `mz_objects`. This ensures that whenever a dependency is deleted we also
+  implicitly delete any notices that depend on it.
+- `C` corresponds to [all `mz_all_optimizer_notices`
+  entries](#mz_all_optimizer_notices) that that have a `fingerprint` contained
+  [in `mz_ignored_optimizer_notices`](#mz_ignored_optimizer_notices).
+
+#### `mz_all_optimizer_notices`
+
+A new `BuiltinTable` that contains an unfiltered list of all optimizer notices
+emitted by the optimizer when planning materialized views and indexes. The
+schema of that table is identical to `mz_optimizer_notices` and its contents are
+[maintained directly by the adapter](#notice-generation-and-management).
+
+#### `mz_ignored_optimizer_notices`
+
+A new `BuiltinSource` used to keep track of ignored notices. Because of the
+`fingerprint` characteristics ignored notices will continue to be ignored even
+if we re-create the objects associated via `object_id` and `dependency_ids` as
+long as we use identical DDL statements.
+
+### Notice Generation and Management
+
+#### Notice Generation
+
+An optimizer pass produces a vector of notices as a side effect of an
+optimization run. In order to persist those notices in
+`mz_all_optimizer_notices` we need to be able to emit retractions for a
+previously inserted notice (for example when we delete the associated catalog
+item). However, the notice structure of the notices generated by the optimizer
+is not suitable for emitting retractions - for example:
+
+1. It doesn't have an optional `created_at` field.
+2. Some of the text fields need to be resolved against catalog items that might
+   be renamed or removed in the meantime.
+
+In order to protect ourselves against invalid retraction issues arising from
+these problems we make a distinction between the notice type produced by the
+optimizer (`RawOptimizerNotice`) and the notice type stored in the
+`CatalogPlans` struct (`OptimizerNotice`).
+
+Further, the type of the notices stored in `DataflowMeatinfo::optimizer_notices`
+depends on the execution stage - within the optimizer we still need to use
+`RawOptimizerNotice`, but before storing the final result in the `CatalogPlans`
+struct we need to move to a `OptimizerNotice` vector (done with a
+`CatalogState::render_notices` call). We use the `CatalogPlans` entries to
+generate insertions and deletions in the `mz_all_optimizer_notices` table.
+
+Also, since we expect to have more than two notice types in the long term, we
+reorganize the code a bit so adding a new notice is easier in the future. The
+required steps are as follows:
+
+1. Create a `mz_transform::notice::<notice_type>` submodule.
+2. Define a struct for the new notice in that module.
+3. Implement `OptimizerNoticeApi` for that struct.
+4. Re-export the notice type in `mz_transform::notice`.
+5. Add the notice type to the `raw_optimizer_notices` macro which generates the
+   `RawOptimizerNotice` enum and other boilerplate code.
+
+#### `mz_all_optimizer_notices` maintenance
```

Comment:
Do we have an estimate to how many entries might exist in `mz_all_optimizer_notices`? We should double check that filter pushdown works on a column, preferably on the `object_id` column which I'm guessing is what users will mostly filter on